File: python/examples/settlement_workflow.py
from __future__ import annotations

import os

from chancery_reference.client import (
    ChanceryClient,
    SettlementOperationRequest,
    SettlementTransactionRequest,
)
from chancery_reference.solana_transaction import load_solana_keypair_file


def required_environment_variable(name: str) -> str:
    value = os.environ.get(name)
    if value is None or value == "":
        raise ValueError(f"Missing environment variable {name}")
    return value


def parse_action(value: str) -> str:
    if value not in ("mint", "redeem"):
        raise ValueError("ACTION must be mint or redeem")
    return value


def parse_mode(value: str) -> str:
    if value not in ("direct", "delegated", "trilateral"):
        raise ValueError("MODE must be direct, delegated, or trilateral")
    return value


def parse_unsigned_integer(value: str, name: str) -> int:
    if not value.isdecimal():
        raise ValueError(f"{name} must be a raw unsigned integer")
    return int(value)


def comma_separated_values(value: str | None) -> tuple[str, ...]:
    if value is None or value.strip() == "":
        return ()
    return tuple(part.strip() for part in value.split(",") if part.strip() != "")


def main() -> int:
    rpc_endpoint = required_environment_variable("RPC_URL")
    client = ChanceryClient(rpc_endpoint, "confirmed")

    action = parse_action(required_environment_variable("ACTION"))
    mode = parse_mode(os.environ.get("MODE", "direct"))
    asset_mint = required_environment_variable("ASSET_MINT")
    principal = required_environment_variable("PRINCIPAL")
    signer_paths = comma_separated_values(required_environment_variable("KEYPAIR_PATHS"))
    keypairs = tuple(load_solana_keypair_file(path) for path in signer_paths)
    if len(keypairs) == 0:
        raise ValueError("KEYPAIR_PATHS did not contain a keypair path")

    request = SettlementOperationRequest(
        action=action,
        mode=mode,
        asset_mint=asset_mint,
        principal=principal,
        amount=parse_unsigned_integer(required_environment_variable("AMOUNT"), "AMOUNT")
        if mode == "direct"
        else None,
        minimum_output=parse_unsigned_integer(os.environ["MINIMUM_OUTPUT"], "MINIMUM_OUTPUT")
        if "MINIMUM_OUTPUT" in os.environ
        else None,
        pathway_id=os.environ.get("PATHWAY_ID"),
        intent_id=required_environment_variable("INTENT_ID") if mode != "direct" else None,
        executor=os.environ.get("EXECUTOR"),
        principal_b=os.environ.get("PRINCIPAL_B"),
        source_token_account=os.environ.get("SOURCE_TOKEN_ACCOUNT"),
        destination_token_account=os.environ.get("DESTINATION_TOKEN_ACCOUNT"),
        fee_recipient_token_account=os.environ.get("FEE_RECIPIENT_TOKEN_ACCOUNT"),
    )

    inspection = client.inspect(request)
    print(ChanceryClient.stringify(inspection))
    if not inspection.ready:
        return 2

    lookup_table_addresses = comma_separated_values(os.environ.get("LOOKUP_TABLES"))
    transaction_request = SettlementTransactionRequest(
        fee_payer=os.environ.get("FEE_PAYER", keypairs[0].public_key),
        keypairs=keypairs,
        address_lookup_table_addresses=lookup_table_addresses,
        commitment="confirmed",
    )
    simulation = client.simulate_transaction(inspection, transaction_request)
    print(ChanceryClient.stringify(simulation))
    if os.environ.get("SUBMIT") == "true" and simulation.simulation.error is None:
        submitted = client.submit_transaction(inspection, transaction_request)
        print(ChanceryClient.stringify(submitted))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/examples/SettlementWorkflow.ts
import {
    ChanceryClient,
    loadSolanaKeypairFile,
    type SettlementOperationRequest,
    type SettlementTransactionRequest,
    type SettlementAction,
    type SettlementMode,
} from "../src/index.js";

function requiredEnvironmentVariable(name: string): string {
    const value = process.env[name];
    if (value === undefined || value.length === 0) {
        throw new Error(`Missing environment variable ${name}`);
    }
    return value;
}

function parseAction(value: string): SettlementAction {
    if (value === "mint" || value === "redeem") {
        return value;
    }
    throw new Error("ACTION must be mint or redeem");
}

function parseMode(value: string): SettlementMode {
    if (value === "direct" || value === "delegated" || value === "trilateral") {
        return value;
    }
    throw new Error("MODE must be direct, delegated, or trilateral");
}

function parseUnsignedInteger(value: string, name: string): bigint {
    if (!/^[0-9]+$/.test(value)) {
        throw new Error(`${name} must be a raw unsigned integer`);
    }
    return BigInt(value);
}

function commaSeparatedValues(value: string | undefined): readonly string[] {
    if (value === undefined || value.trim().length === 0) {
        return [];
    }
    return value.split(",").map((part) => part.trim()).filter((part) => part.length > 0);
}

async function main(): Promise<void> {
    const rpcEndpoint = requiredEnvironmentVariable("RPC_URL");
    const chanceryClient = new ChanceryClient(rpcEndpoint, "confirmed");

    const action = parseAction(requiredEnvironmentVariable("ACTION"));
    const mode = parseMode(process.env.MODE ?? "direct");
    const assetMint = requiredEnvironmentVariable("ASSET_MINT");
    const principal = requiredEnvironmentVariable("PRINCIPAL");
    const signerPaths = commaSeparatedValues(requiredEnvironmentVariable("KEYPAIR_PATHS"));
    const keypairs = signerPaths.map((path) => loadSolanaKeypairFile(path));
    const firstKeypair = keypairs[0];
    if (firstKeypair === undefined) {
        throw new Error("KEYPAIR_PATHS did not contain a keypair path");
    }

    const commonRequest = {
        action,
        mode,
        assetMint,
        principal,
        ...(process.env.PATHWAY_ID === undefined ? {} : { pathwayId: process.env.PATHWAY_ID }),
        ...(process.env.MINIMUM_OUTPUT === undefined
            ? {}
            : { minimumOutput: parseUnsignedInteger(process.env.MINIMUM_OUTPUT, "MINIMUM_OUTPUT") }),
        ...(process.env.EXECUTOR === undefined ? {} : { executor: process.env.EXECUTOR }),
        ...(process.env.PRINCIPAL_B === undefined ? {} : { principalB: process.env.PRINCIPAL_B }),
        ...(process.env.SOURCE_TOKEN_ACCOUNT === undefined
            ? {}
            : { sourceTokenAccount: process.env.SOURCE_TOKEN_ACCOUNT }),
        ...(process.env.DESTINATION_TOKEN_ACCOUNT === undefined
            ? {}
            : { destinationTokenAccount: process.env.DESTINATION_TOKEN_ACCOUNT }),
        ...(process.env.FEE_RECIPIENT_TOKEN_ACCOUNT === undefined
            ? {}
            : { feeRecipientTokenAccount: process.env.FEE_RECIPIENT_TOKEN_ACCOUNT }),
    };
    const request: SettlementOperationRequest = mode === "direct"
        ? {
            ...commonRequest,
            amount: parseUnsignedInteger(requiredEnvironmentVariable("AMOUNT"), "AMOUNT"),
        }
        : {
            ...commonRequest,
            intentId: requiredEnvironmentVariable("INTENT_ID"),
        };

    const inspection = await chanceryClient.inspect(request);
    process.stdout.write(`${ChanceryClient.stringify(inspection)}\n`);
    if (!inspection.ready) {
        process.exitCode = 2;
        return;
    }

    const lookupTableAddresses = commaSeparatedValues(process.env.LOOKUP_TABLES);
    const transactionRequest: SettlementTransactionRequest = {
        feePayer: process.env.FEE_PAYER ?? firstKeypair.publicKey,
        keypairs,
        commitment: "confirmed",
        ...(lookupTableAddresses.length === 0 ? {} : { addressLookupTableAddresses: lookupTableAddresses }),
    };
    const simulation = await chanceryClient.simulateTransaction(inspection, transactionRequest);
    process.stdout.write(`${ChanceryClient.stringify(simulation)}\n`);

    if (process.env.SUBMIT === "true" && simulation.simulation.err === null) {
        const submitted = await chanceryClient.submitTransaction(inspection, transactionRequest);
        process.stdout.write(`${ChanceryClient.stringify(submitted)}\n`);
    }
}

main().catch((error: unknown) => {
    const message = error instanceof Error ? error.stack ?? error.message : String(error);
    process.stderr.write(`${message}\n`);
    process.exitCode = 1;
});
